example-apps/token_curated_registry/types/registry.go

```diff
@@ -0,0 +1,35 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type Listing struct {
+	Identifier string
+	Votes      int64
+}
+
+// Create new Voter for address on each Listing
+type Voter struct {
+	Owner      sdk.Address
+	Identifier string
+}
+
+// Vote revealed during reveal phase
+type Vote struct {
+	Choice bool
+	Power  int64
+}
+
+type Ballot struct {
```

Comment:
What's a Ballot exactly? It's hard for me by just looking at the code to figure out what Listings, Ballot are. Here is how I understand the process:

1. I apply for a listing. A temporary listing is created for people to challenge me until APPLY_STAGE_LEN is over or I win a challenge. Is there a type for this "temporary listing"?
2. If I get challenged at any point, a vote opens. Is that "Ballot"?
3. If I don't get challenged or win the challenge, my "Temporary listing" becomes a "Listing", is that correct?

Btw, is there a way for candidates to add information to their Listings to prove that they deserve to be included? Maybe a "Description" or "Details" field in the "Listing" struct

Reply:
Hello, thank you for the great feedback. I will add some comments and documentation to make this more clear.

When you apply for a listing an inactive Ballot is created and added to the ballot store.

If no one challenges you and the apply period passes, a Listing object is created and added to the registry store.

If the listing is challenged before end of apply period, ballot gets activated. It then goes through a commit-reveal phase. During the reveal phase the ballot collects all the valid votes. If you win the challenge, the Listing object is created and added to registry store.

The Ballot is persisted even after the challenge ends because I currently use it to hand out rewards. Perhaps, there is a better way to do this.

Including Details in Listing is a great idea.

Reply:
But as I understand the design (from medium article), a ballot (i.e. a challenge) can be created even after the Listing has passed the apply period or after having been already challenged. This is the case if the candidate lost right to be listed over time for example.

Why not have two different maps of`Listings` in the store, one `Pending` and one `Active` or `Official`. A challenge can be created on both, this does not change. The only difference is that Listings are moved from `Pending` to `Active` if they remain in `Pending` long enough without being challenged.